File: 0112-path-sum/0112-path-sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func hasPathSum(root *TreeNode, targetSum int) bool {
    if root == nil {
        return false
    }
    
    remainSum := targetSum - root.Val
    
    if root.Left == nil && root.Right == nil {
        return remainSum == 0
    }
    
    return hasPathSum(root.Left, remainSum) || hasPathSum(root.Right, remainSum)
}

File: 0112-path-sum/0112-path-sum.php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {

    /**
     * @param TreeNode $root
     * @param Integer $targetSum
     * @return Boolean
     */
    function hasPathSum($root, $targetSum) {
        return $this->DFS($root, 0, $targetSum);
    }
    
    function DFS($node, $sum, $targetSum) {
        if(!$node) {
            return false;
        }
        
        $sum += $node->val;
        
        if(!$node->left && !$node->right){
            return $sum === $targetSum;
        }
        
        return $this->DFS($node->left, $sum, $targetSum) || $this->DFS($node->right, $sum, $targetSum);
    }
}